mono/metadata/threads.c

```diff
@@ -1650,40 +1655,60 @@ mono_thread_set_name_internal (MonoInternalThread *this_obj, MonoString *name, g
 		mono_error_set_invalid_operation (error, "Thread.Name can only be set once.");
 		return;
 	}
-	if (this_obj->name) {
-		g_free (this_obj->name);
-		this_obj->name_len = 0;
-	}
-	if (name) {
-		this_obj->name = g_memdup (mono_string_chars (name), mono_string_length (name) * sizeof (gunichar2));
-		this_obj->name_len = mono_string_length (name);
 
+	g_free (this_obj->name);
+	this_obj->name = 0;
+	this_obj->name_len = 0;
+
+	if (length > 0) {
+		this_obj->name = utf16 ? (gunichar2*)g_memdup (utf16, length * sizeof (gunichar2))
+							   : g_utf8_to_utf16 (utf8, length, NULL, NULL, NULL);
+		this_obj->name_len = length;
 		if (permanent)
 			this_obj->flags |= MONO_THREAD_FLAG_NAME_SET;
 	}
-	else
-		this_obj->name = NULL;
 
 	if (!(this_obj->state & ThreadState_Stopped))
 		tid = thread_get_tid (this_obj);
 
 	UNLOCK_THREAD (this_obj);
 
 	if (this_obj->name && tid) {
-		char *tname = mono_string_to_utf8_checked (name, error);
-		return_if_nok (error);
+		const char *tname = 0;
+		if (utf8)
+			tname = utf8;
+		else {
+			tname = mono_utf16_to_utf8 (utf16, length, error);
+			return_if_nok (error);
+		}
 		MONO_PROFILER_RAISE (thread_name, ((uintptr_t)tid, tname));
 		mono_native_thread_set_name (tid, tname);
-		mono_free (tname);
+		if (!utf8)
+			mono_free (MONO_CONST_CAST(char*)(tname));
 	}
 }
 
+void
+mono_thread_set_name_cstr_internal (MonoInternalThread *thread, const char* name, gboolean permanent, gboolean reset, MonoError *error)
+{
+	mono_thread_set_name_utf_internal (thread, strlen(name), name, NULL, permanent, reset, error);
```

Comment:
On second thought, I'm not feeling so good about `mono_thread_set_name_cstr_internal`.  The `cstr` is needlessly ambiguous from the caller's perspective, I think.  We should say what we mean: UTF8.

And we should check for conversion errors from `g_utf8_to_utf16` in `mono_thread_set_name_utf_internal`.

Reply:
And here remove the _cstr_ function entirely, and then some dynamic utf8 conversion (could remove it statically), though a bit painful.
https://github.com/jaykrell/mono/commit/38bc23b8d43fdf7c21b908bf198076824c843b55